File: Ruby/stack/stack.rb
class StackNode
	attr_accessor :data, :next
	
	def initialize(data)
		@data = data
		@next = nil
	end
end

class Stack
	def initialize(elems = nil)
		@top = nil
		@size = 0
		if elems
			elems.each { |e| push(e) }
		end
	end
	
	def length
		@size
	end
	
	def empty?
		@size == 0
	end
	
	def push(e)
		if @size == 0
			@top = StackNode.new(e)
		else
			temp = SinglyNode.new(e)
			temp.next = @top
			@top = temp
		end
		@size += 1
	end
	
	def peek
		@top&.data
	end
	
	def pop
		data = @top&.data
		if data
			@top = @top.next
			@size -= 1
		end
		data
	end
	
	def include?(e)
		current = @top
		while current
			return true if current.data = e
			current = current.next
		end
		false
	end
	
	def clear
		@top = nil
		@size = 0
	end
end

File: C/stack/stack.c
#include <stdio.h>
#include <stdlib.h>
#include "stack.h"

stacknode* stacknode_new(void* data) {
	stacknode* node = NULL;
	node = malloc(sizeof(stacknode));
	node->data = data;
	node->next = NULL;
	return node;
}

stack* stack_new(void) {
	stack* s = NULL;
	s = malloc(sizeof(stack));
	s->top = NULL;
	s->size = 0;
	return s;
}

void stack_init(stack* s) {
	s->top = NULL;
	s->size = 0;
}

void stack_free(stack* s) {
	if (s->size == 0) {
		free(s);
		return;
	}
	stacknode* current = s->top;
	stacknode* temp = current->next;
	while (temp != NULL) {
		free(current);
		current = temp;
		temp = temp->next;
	}
	free(current);
	free(s);
}

int stack_size(stack* s) {
	return s->size;
}

int stack_empty(stack* s) {
	return s->size == 0;
}

void stack_push(stack* s, void* e) {
	stacknode* temp = stacknode_new(e);
	if (s->size == 0) {
		s->top = temp;
	}
	else {
		temp->next = s->top;
		s->top = temp;
	}
	++(s->size);
}

void* stack_peek(stack* s) {
	if (s->top == NULL) {
		return NULL;
	}
	return s->top->data;
}

void* stack_pop(stack* s) {
	if (s->size == 0) {
		return NULL;
	}
	stacknode* node = s->top;
	void* data = node->data;
	s->top = node->next;
	free(node);
	return data;
}

int stack_contains(stack* s, void* e) {
	stacknode* current = s->top;
	while (current != NULL) {
		if (current->data == e) {
			return 1;
		}
		current = current->next;
	}
	return 0;
}

void stack_clear(stack* s) {
	if (s->size == 0) {
		return;
	}
	stacknode* current = s->top;
	stacknode* temp = current->next;
	while (temp != NULL) {
		free(current);
		current = temp;
		temp = temp->next;
	}
	free(current);
	
	s->top = NULL;
	s->size = 0;
}
